src/common/string/filter/sequence-matching.ts

```diff
@@ -64,21 +80,72 @@ export const fuzzySequentialMatch = (
 export interface ScorableTextItem {
   score?: number;
   strings: string[];
+  decoratedStrings?: string[][];
 }
 
 type FuzzyFilterSort = <T extends ScorableTextItem>(
   filter: string,
-  items: T[]
+  items: T[],
+  decorate?: MatchDecorator
 ) => T[];
 
-export const fuzzyFilterSort: FuzzyFilterSort = (filter, items) => {
+export const fuzzyFilterSort: FuzzyFilterSort = (
+  filter,
+  items,
+  decorate = createMatchDecorator("[", "]")
+) => {
   return items
     .map((item) => {
-      item.score = fuzzySequentialMatch(filter, item);
+      const match = fuzzySequentialMatch(filter, item, decorate);
+
+      item.score = match?.score;
+      item.decoratedStrings = match?.decoratedStrings;
+
       return item;
     })
     .filter((item) => item.score !== undefined)
     .sort(({ score: scoreA = 0 }, { score: scoreB = 0 }) =>
       scoreA > scoreB ? -1 : scoreA < scoreB ? 1 : 0
     );
 };
+
+type MatchDecorator = (word: string, scores?: FuzzyScore) => string[];
+export const createMatchDecorator: (
+  left: string,
+  right: string
+) => MatchDecorator = (left, right) => (word, scores) =>
+  _decorateMatch(word, [left, right], scores);
+
+const _decorateMatch: (
+  word: string,
+  surroundWith: [string, string],
```

Comment:
```suggestion
  surroundWith: [templateResult, templateResult],
```

Reply:
How about `string | TemplateResult`?

Reasons being that:
1. The test throws an exception if I try to import `html` into either `sequence_matching.test.ts` or `sequence-matching.ts`
2. Allowing `string` makes it easier to write unit tests for highlighting, and allows for debugging to console log
3. Future components that use this filter may not want to inject html (can't imagine why not tho)

Reply:
Oh wait... now I remember why I went with `string`. The way text decoration works is we break apart the string into the unmatched and matched chunks, then use string concatenation to glue it all back together along with our decorations:

See these lines: https://github.com/home-assistant/frontend/blob/d1b512e5a2006a46d3ee69fe5307e3db41b41a4b/src/common/string/filter/sequence-matching.ts#L134-L141

How do we concatenate `left` and `right` to the string if they are of type `TemplateResult` without using `unsafeHtml`?

Reply:
No need to concatenate a TemplateResult, just put it in an array, and lit will do it for you :-)

Reply:
> But then later we wouldn't be able to split on the "::" delimiter to restore the category key and category text fragments.

I'm not too happy about that anyway but have not thought of a better way too yet...

Reply:
> No need to concatenate a TemplateResult, just put it in an array, and lit will do it for you :-)

What! Well... that's pretty cool. Not sure it helps with the "::" issue though. Which, yeah I'm not too happy about either, but all the other solutions I came up with were much more complex.

One possibility might be to add an option in the fuzzy matcher to treat an array of separate strings as if they were a single string. That way, the matcher at least has knowledge of where the string separations should be and could maybe use that during the match algo somehow to deconstruct the match properly.

Reply:
>Well... that's pretty cool. Not sure it helps with the "::" issue though.

Maybe instead of `string.split("::")` I could instead do some kind of `findDelimiter(wordChunkArray, "::")` which looks into the innards of each TemplateResult, finds the delimiter, then splits _splices_ the wordChunkArray at that element.